docs/changes/16181.other.rst

```diff
@@ -0,0 +1,11 @@
+Public APIs that expose a ``copy`` argument and that previously set ``False``
+as a default value now use ``None`` instead if Numpy 2.0 or newer is installed.
+Default behaviour is unaffected, however explicitly passing ``copy=False`` will
+now raise an error if copying cannot be avoided.
```

Comment:
> A logical question here is whether we should include how to adjust it. Should we link to astropy.utils.compat.COPY_IF_NEEDED?

I had thought the right answer is to not pass a `copy=...` argument at all, deferring to the default. But this may be upsetting to people that want to be explicit and not rely on a default.

I had also wondered about `copy=None`. The table says that the behavior is formally undefined but in practice for numpy 1.x it is the same as `False`.

Reply:
I'll adjust the note, and I agree it belongs in the what's new.

> A logical question here is whether we should include how to adjust it. Should we link to `astropy.utils.compat.COPY_IF_NEEDED`?

As far as I understand, this module isn't considered public API (we just removed `PYTHON_LT_*` constants from it without a warning), so I wouldn't want to encourage downstream to rely on it. That said, we could recommend copying its implementation (the tricky part, I think, is just in how to properly check numpy's version at runtime using `packaging.version`). What do you think ?

> A more pragmatic question is whether we need tests to check that things do not percolate badly.

Could you clarify what you mean here ? From my standpoint, the "percolating badly" scenario would be #16174 😅 .

Reply:
Oops, going on a race condition with @taldcroft .

> I had thought the right answer is to not pass a `copy=...` argument at all, deferring to the default. But this may be upsetting to people that want to be explicit and not rely on a default.

Well it *is* documented as mostly for internal usage, and people who want to be explicit about *not* copying have no recourse to do so in numpy 1, so I'm not too worried.

> I had also wondered about `copy=None`. The table says that the behavior is formally undefined but in practice for numpy 1.x it is the same as False.

_unless you're using a type checker_, but essentially yes ! We could even recommend using a plain `None` as a backward compatible and cheap replacement.

Reply:
I don't think one can just use `copy=None` on all numpy versions:
```
In [2]: np.__version__
Out[2]: '1.24.2'

In [3]: np.array([1, 2, 3], copy=None)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [3], line 1
----> 1 np.array([1, 2, 3], copy=None)

ValueError: NoneType copy mode not allowed.
```

This is also what I meant by tests. Since we use `copy=False` basically only internally, I'm not so sure that we cover most cases users might do. I *think* we should be OK, though, since in most cases one would not be asking for a dtype change, so, say, `SkyCoord(ra_array, dec_array, copy=False)` should just continue to work.

Anyway, bottom line to me seems that we should just try to make sure this is out when numpy 2.0 comes out, so that we can get reports from users. Indeed, perhaps this is the one time that we should ask people to run their test suites on a point release...

Reply:
> I don't think one can just use copy=None on all numpy versions:

Oh gosh, how did I not check my assumptions... I'll edit the PR description, thanks !

> perhaps this is the one time that we should ask people to run their test suites on a point release...

just to be clear, you mean astropy 6.1.0 here (not 6.0.1), right ?

Reply:
I included Marten's suggestion to rephrase the changlog, and I've removed ``Table`` from the list. I actually don't know why I included it at first, probably a mistake !
I'll work on the what's new in the morning so hopefully this will be ready to review again tomorrow !

Reply:
> just to be clear, you mean astropy 6.1.0 here (not 6.0.1), right ?

Depends on whether we'll still do a 6.0.x with numpy 2.0 support. Haven't we been backporting most things for that? (At least, things like new ufuncs, etc.) I think we need to have *some* public astropy version that works with numpy 2.0 when that comes out...

Reply:
Yes, 6.0.x should be compatible already and hopefully won't require more changes, but we haven't backported this PR's family of patches, which is why I got confused. Anyway, the window between 2.0.0rc1 and final should be long enough for us to land a compatible release.

Reply:
Ah, yes, I guess we should backport those PRs that ensure `6.0.x` passes on numpy 2.0...

Reply:
It'll probably be easier to do in bulk. I suggest we move this discussion to #16167

Reply:
Actually I think there are conflicting desires here: if we're adding a what's new, then backporting (along with supporting PRs) becomes impossible (patch releases don't have what's news).
So, the easiest way to resolve this tension would probably be to not have a what's new, and instead make the changelog entry even longer. As long as we make sure that the release team is aware of the importance of this patch and can properly emphasize it as part of our numpy 2.0 compatibility layer, it should be good ?

Reply:
Yes, makes sense!

Reply:
I really think we should only release all these changes in 6.1.0 and have a What's New.

If there is a chance that numpy 2.0 will get released before early May then we should move the astropy release forward. Having these changes in a bug fix release is (IMO) totally incompatible with our policy for such releases. Despite a lot of effort and review, I think it is credible that the numpy 2.0 compatibility changes will break some existing code somewhere. Users don't expect this for a bug fix release.

Reply:
OK, that makes sense too. So, we basically will make clear that astropy 6.0 requires numpy < 2.

I had vaguely considered the idea of an early 6.1 as well - seems worth a new issue! #16200

Reply:
ok then I'll make a what's new from the changelog